File: hub/app/api/encounter.py
# ============================================================
# As Above, So Below. As Within, So Without.
# The Future Dictates the Past and the Past is Always Present.
# ============================================================

"""Encounter Generator API — LLM + deterministic XP budget."""

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field

from app.agents.encounter_builder import EncounterBuilderAgent, BuildResult
from app.llm.ollama_client import OllamaClient
from app.config import get_settings
from app.db.repository import CampaignRepository
from app.rag.retriever import Retriever
from app.api.deps import get_repo

router = APIRouter(prefix="/encounter", tags=["encounter"])


class EncounterRequest(BaseModel):
    party_level: int = Field(..., ge=1, le=20)
    party_size: int = Field(..., ge=1, le=10)
    threat: str = Field(..., pattern="^(trivial|low|moderate|severe|extreme)$")
    theme: str = Field(..., min_length=1, max_length=200)


class EncounterMonsterResponse(BaseModel):
    name: str
    count: int
    level: int
    xp_each: int
    total_xp: int
    source_book: str


class EncounterResponse(BaseModel):
    target_xp: int
    total_xp: int
    party_level: int
    party_size: int
    threat: str
    theme: str
    monsters: list[EncounterMonsterResponse]


@router.post("/generate", response_model=EncounterResponse)
async def generate_encounter(
    request: EncounterRequest,
    repo: CampaignRepository = Depends(get_repo),
) -> EncounterResponse:
    settings = get_settings()
    llm = OllamaClient(settings.ollama_host)
    retriever = Retriever(repo)
    agent = EncounterBuilderAgent(llm, retriever)

    try:
        result: BuildResult = await agent.build(
            party_level=request.party_level,
            party_size=request.party_size,
            threat=request.threat,
            theme=request.theme,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    finally:
        await llm.close()

    return EncounterResponse(
        target_xp=result.target_xp,
        total_xp=result.total_xp,
        party_level=result.party_level,
        party_size=result.party_size,
        threat=result.threat,
        theme=result.theme,
        monsters=[
            EncounterMonsterResponse(
                name=m.name,
                count=m.count,
                level=m.level,
                xp_each=m.xp_each,
                total_xp=m.total_xp,
                source_book=m.source_book,
            )
            for m in result.monsters
        ],
    )

File: spoke/lib/models/encounter.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

import 'dart:convert';

/// A monster in a generated encounter.
class EncounterMonster {
  final String name;
  final int count;
  final int level;
  final int xpEach;
  final int totalXp;
  final String sourceBook;
  final String content;

  const EncounterMonster({
    required this.name,
    required this.count,
    required this.level,
    required this.xpEach,
    required this.totalXp,
    required this.sourceBook,
    required this.content,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'count': count,
        'level': level,
        'xp_each': xpEach,
        'total_xp': totalXp,
        'source_book': sourceBook,
        'content': content,
      };

  factory EncounterMonster.fromJson(Map<String, dynamic> json) =>
      EncounterMonster(
        name: json['name'] as String,
        count: json['count'] as int? ?? 1,
        level: json['level'] as int? ?? 1,
        xpEach: json['xp_each'] as int? ?? 0,
        totalXp: json['total_xp'] as int? ?? 0,
        sourceBook: json['source_book'] as String? ?? '',
        content: json['content'] as String? ?? '',
      );

  EncounterMonster copyWith({
    String? name,
    int? count,
    int? level,
    int? xpEach,
    int? totalXp,
    String? sourceBook,
    String? content,
  }) =>
      EncounterMonster(
        name: name ?? this.name,
        count: count ?? this.count,
        level: level ?? this.level,
        xpEach: xpEach ?? this.xpEach,
        totalXp: totalXp ?? this.totalXp,
        sourceBook: sourceBook ?? this.sourceBook,
        content: content ?? this.content,
      );
}

/// The full generated encounter result.
class GeneratedEncounter {
  final int targetXp;
  final int totalXp;
  final int partyLevel;
  final int partySize;
  final String threat;
  final String theme;
  final List<EncounterMonster> monsters;

  const GeneratedEncounter({
    required this.targetXp,
    required this.totalXp,
    required this.partyLevel,
    required this.partySize,
    required this.threat,
    required this.theme,
    required this.monsters,
  });

  Map<String, dynamic> toJson() => {
        'target_xp': targetXp,
        'total_xp': totalXp,
        'party_level': partyLevel,
        'party_size': partySize,
        'threat': threat,
        'theme': theme,
        'monsters': monsters.map((m) => m.toJson()).toList(),
      };

  factory GeneratedEncounter.fromJson(Map<String, dynamic> json) =>
      GeneratedEncounter(
        targetXp: json['target_xp'] as int? ?? 0,
        totalXp: json['total_xp'] as int? ?? 0,
        partyLevel: json['party_level'] as int? ?? 1,
        partySize: json['party_size'] as int? ?? 4,
        threat: json['threat'] as String? ?? 'moderate',
        theme: json['theme'] as String? ?? '',
        monsters: (json['monsters'] as List<dynamic>? ?? [])
            .map((e) => EncounterMonster.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  int get monsterCount => monsters.fold(0, (sum, m) => sum + m.count);

  @override
  String toString() =>
      'GeneratedEncounter(partyLevel: $partyLevel, threat: $threat, theme: $theme, '
      'targetXp: $targetXp, totalXp: $totalXp, monsters: $monsterCount)';
}